packages/koa-react-sidecar/src/render/render.tsx

```diff
@@ -1,46 +1,106 @@
+import {Omit} from '@shopify/useful-types';
 import {Header} from '@shopify/react-network';
+import {getAssets} from '@shopify/sewing-kit-koa';
+import {createGraphQLClient, GraphQLClientOptions} from '../graphql-client';
+import {getLogger} from '../logger';
+
+export type RenderContext = Context & {
+  graphQLClientOptions?: GraphQLClientOptions;
+  graphQLClient?: ApolloClient<{}>;
+};
 
-export type RenderContext = Context & {};
+export interface CreateRenderOptions {
+  render: (ctx: RenderContext) => React.ReactElement<any>;
+  graphQLClientOptions?: Omit<GraphQLClientOptions, 'host'>;
+}
 
-const logger = console;
+export function createRender(options: CreateRenderOptions) {
+  const {render, graphQLClientOptions} = options;
 
-export function createRender(
-  renderFunction: (ctx: RenderContext) => React.ReactElement<any>,
-) {
-  return async function render(ctx: RenderContext) {
-    const app = renderFunction(ctx);
+  return async function renderFunction(ctx: RenderContext) {
+    const logger = getLogger(ctx);
+    const assets = getAssets(ctx);
     const networkManager = new NetworkManager();
     const htmlManager = new HtmlManager();
+    const graphQLClient =
+      graphQLClientOptions &&
+      createGraphQLClient({host: ctx.host, ...graphQLClientOptions});
+    const ApolloBridge = createApolloBridge();
+    const asyncAssetManager = new AsyncAssetManager();
+    const app = render({...ctx, graphQLClientOptions, graphQLClient});
 
     try {
       await extract(app, {
         decorate(app) {
           return (
-            <NetworkContext.Provider value={networkManager}>
-              <HtmlContext.Provider value={htmlManager}>
-                {app}
-              </HtmlContext.Provider>
-            </NetworkContext.Provider>
+            <HtmlContext.Provider value={htmlManager}>
+              <AsyncAssetContext.Provider value={asyncAssetManager}>
+                <NetworkContext.Provider value={networkManager}>
+                  <ApolloBridge>{app}</ApolloBridge>
+                </NetworkContext.Provider>
+              </AsyncAssetContext.Provider>
+            </HtmlContext.Provider>
           );
         },
+        afterEachPass({renderDuration, resolveDuration, index, finished}) {
+          const pass = `Pass number ${index} ${
+            finished ? ' (this was the final pass)' : ''
+          }`;
+          const rendering = `Rendering ${renderDuration}ms`;
+          const resolving = `Resolving ${resolveDuration}ms`;
+
+          logger.log(pass);
+          logger.log(rendering);
+          logger.log(resolving);
+        },
       });
     } catch (error) {
       logger.error(error);
       throw error;
     }
 
     applyToContext(ctx, networkManager);
-    const response = renderToString(<Html manager={htmlManager}>{app}</Html>);
+
+    const [styles, scripts] = await Promise.all([
+      assets.styles(),
+      assets.scripts(),
+    ]);
+
+    const serializedInitialApolloData = graphQLClient ? (
```

Comment:
This is more of what I mean, we handle setting up a bunch of stuff for them here but it feels awkward that they need to grab a bunch of very specific libraries to use that and know about exactly how we've set up this serialization. With `extract` and the self-serializer pattern, everything else their app wants to setup they can just do with their own components, I'm wondering why this should be different. I have two possible directions I'd prefer to see this go:

- Build a self-serializer for Apollo clients / data and just let the user use that if they want. Then this library continues to only care about being a minimal setup for `react-html` apps.

- Make this even more opinionated about apollo stuff. Render the provider for them, maybe provide a simple client api that gives them access to the serialized data.

Personally, after sleeping on it and looking over this I think I prefer the first approach. Encouraging self-serializers and keeping this lightweight / leveraging `@shopify/react-html`'s superpowers of letting apps handle this stuff within their universal code seems like a big win for needing to update this thing less.